cluster/gce/delete-stranded-ingress-resources.sh

```diff
@@ -0,0 +1,133 @@
+}
+
+v_gcloud() {
+	echo "+ gcloud --project=${PROJECT}" "$@"
+	(
+		set +e
+		command gcloud --project="${PROJECT}" "$@"
+	)
+}
+
+has_health_status() {
+	local backend
+	backend="$1"
+
+	local out
+	out="$(gcloud compute backend-services get-health "${backend}" \
+		--global --format='value(status.healthStatus)')"
+	[[ -n "$out" ]]
+}
+
+cleanup_resources() {
+	local backend backend_id
+	backend="$1"
+	backend_id="$(echo "${backend}" | grep -Eo '\-\-([0-9a-f]+)' /dev/stdin | sed 's/^--//')"
+
+	local health_check firewall url_map target_http_proxy target_https_proxy \
+		forwarding_rule instance_groups
+
+	firewall="k8s-fw-l7--${backend_id}"
+
+	url_map="$(gcloud compute url-maps list --format='value(name)' \
+		--filter=defaultService~"${backend}")"
+	if [[ -n "${url_map}" ]]; then
+		forwarding_rule="${url_map/-um-/-fw-}"
+		target_http_proxy="$(gcloud compute target-http-proxies list	\
+			--format='value(name)' --filter=urlMap~"${url_map}")"
+		target_https_proxy="$(gcloud compute target-https-proxies list \
+			--format='value(name)' --filter=urlMap~"${url_map}")"
+	fi
+
+	health_check="$(gcloud compute backend-services describe \
+		--global "${backend}" --format='value(healthChecks)')"
+
+	instance_groups=$(gcloud compute backend-services describe \
+		--global "${backend}" --format='value(backends.group)' | tr \; '\n')
+
+	log "    FIREWALL=${firewall}"
+	log "    HEALTH_CHECK=${health_check}"
+	log "    URL_MAP=${url_map}"
+	log "    TARGET_HTTP_PROXY=${target_http_proxy}"
+	log "    TARGET_HTTPS_PROXY=${target_https_proxy}"
+	log "    FORWARDING_RULE=${forwarding_rule}"
+	if [[ -n "${instance_groups}" ]]; then
+		log "    INSTANCE_GROUPS:"
+		for ig in ${instance_groups}; do
+			log "    - ${ig}"
+		done
+	fi
+
+	read -p "Clean these up? (y/N?) " -r
+	echo
+	if [[ $REPLY =~ ^[Yy]$ ]]
+	then
+		if [[ -n "${forwarding_rule}" ]]; then
+			v_gcloud compute forwarding-rules delete --global "${forwarding_rule}" -q || true
+		fi
+		if [[ -n "${target_http_proxy}" ]]; then
+			v_gcloud compute target-http-proxies delete "${target_http_proxy}" -q || true
+		fi
+		if [[ -n "${target_https_proxy}" ]]; then
+			v_gcloud compute target-https-proxies delete "${target_https_proxy}" -q || true
+		fi
+		if [[ -n "${url_map}" ]]; then
+			v_gcloud compute url-maps delete "${url_map}" -q || true
+		fi
+		v_gcloud compute backend-services delete --global "${backend}" -q || true
+		v_gcloud compute firewall-rules delete "${firewall}" -q || true
+		v_gcloud compute health-checks delete "${health_check}" -q || true
+		if [[ -n "${instance_groups}" ]]; then
+			for ig in ${instance_groups}; do
+				v_gcloud compute instance-groups unmanaged delete -q "${ig}" || true
+			done
+		fi
+	fi
+
+}
+
+main(){
+	BACKEND_SERVICES=$(gcloud compute backend-services list --global \
+		--format='value(name)' --filter='name ~ ^k8s-be-')
+	log "Found $(echo -n "${BACKEND_SERVICES}" | wc -l) total global backends."
+	for be in ${BACKEND_SERVICES}; do
```

Comment:
Why using backend service as the entry point? There could be cases backend service was cleaned up but some other resources weren't.

Reply:
The script mostly assumes you didn't clean anything up, so the discovery starts from the backend-services.

Another option could be listing resources (health-check, backend-service, forwarding-rule, ...) by their name (e.g. `k8s-be-*`) and seeing what they're attached to is unused, too. e.g.:

```
foreach be in backend-service that starts with k8s-be-*
    delete be, if its health status is empty

foreach fw in firewall
   backend = "s/-fw-/-be-/"
   delete $fw if $backend doesn't exist anymore

foreach hc in healtchecks like k8-be-*
    backend name is the same as healthcheck
    delete $hc if $backend doesn't exist anymore

foreach um in urlmaps like k8s-um-*
    backend is listed on the um's property
    delete $um if its $backend doesn't exist anymore

foreach tp in target-proxies like k8s-tp-*
    backend is listed on the tp's property
    delete $tp if its $backend doesn't exist anymore

foreach fr in forwarding-rules like k8s-fw-*
    target_proxy is listed on the tp's property
    delete $fr if its $target_proxy doesn't exist anymore

...
```

Is this better? (IMO it relies on naming of resources a bit too much.)

Reply:
@MrHohn "getting names from the resource" requires a starting point (in my code it's the "backend-service"). If it's deleted, I have to rely on naming to discover leftover resources? Do you have any good recommendations which route I should follow?